src/llvm/tools/lld/ELF/Arch/MSP430.cpp

```diff
@@ -32,13 +32,38 @@ class MSP430 final : public TargetInfo {
 
 RelExpr MSP430::getRelExpr(RelType Type, const Symbol &S,
                            const uint8_t *Loc) const {
-  return R_ABS;
+  switch (Type) {
+  case R_MSP430_10_PCREL:
+  case R_MSP430_16_PCREL:
+  case R_MSP430_16_PCREL_BYTE:
+  case R_MSP430_2X_PCREL:
+  case R_MSP430_RL_PCREL:
+    return R_PC;
+  default:
+    return R_ABS;
+  }
 }
 
 void MSP430::relocateOne(uint8_t *Loc, RelType Type, uint64_t Val) const {
   switch (Type) {
+  case R_MSP430_8:
+    checkIntUInt(Loc, Val, 8, Type);
+    *Loc = Val;
+    break;
+  case R_MSP430_16:
   case R_MSP430_16_BYTE:
-    write16le(Loc, read16le(Loc) | Val);
+  case R_MSP430_16_PCREL:
+  case R_MSP430_16_PCREL_BYTE:
```

Comment:
I am not a MSP430 expert and given the absence of the correct/full relocations specification,
the below is just an observation atm:

Looking at the bfd code, these seem to be not 100% identical. The difference is in `src_mask` field:
(not sure how it is used)
https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob;f=bfd/elf32-msp430.c;h=2d351d3906cc5f985d43d9b6766ae2b2ab100b90;hb=HEAD#l115

And they are handled a bit differently:
https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob;f=bfd/elf32-msp430.c;h=2d351d3906cc5f985d43d9b6766ae2b2ab100b90;hb=HEAD#l1194